codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java

```diff
@@ -0,0 +1,825 @@
+     * @param status The value for {@link PseudoHeaderName#STATUS}.
+     * @param otherHeaders A an array of key:value pairs. Must not contain any
+     *                     <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.1">pseudo headers</a>
+     *                     or {@code null} names/values.
+     *                     A copy will <strong>NOT</strong> be made of this array. If the contents of this array
+     *                     may be modified externally you are responsible for passing in a copy.
+     * @param validateHeaders {@code true} will run validation on each header name/value pair to ensure protocol
+     *                        compliance.
+     * @return a new read only representation of headers used by servers.
+     */
+    public static ReadOnlyHttp2Headers serverHeaders(AsciiString status, AsciiString[] otherHeaders,
+                                                     boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(new AsciiString[] { PseudoHeaderName.STATUS.value(), status }, otherHeaders,
+                                        validateHeaders);
+    }
+
+    private ReadOnlyHttp2Headers(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders, boolean validateHeaders) {
+        assert (pseudoHeaders.length & 1) == 0; // pseudoHeaders are only set internally so assert should be enough.
+        if ((otherHeaders.length & 1) != 0) {
+            throw newIllegalArgumentException();
+        }
+        if (validateHeaders) {
+            validateHeaders(pseudoHeaders, otherHeaders);
+        }
+        this.pseudoHeaders = pseudoHeaders;
+        this.otherHeaders = otherHeaders;
+    }
+
+    private static IllegalArgumentException newIllegalArgumentException() {
+        return new IllegalArgumentException("pseudoHeaders and otherHeaders must be arrays of [name, value] pairs");
+    }
+
+    private static void validateHeaders(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders) {
+        boolean seenNonPseudoHeader = false;
+        final int pseudoHeadersEnd = pseudoHeaders.length - 1;
+        for (int i = 0; i < pseudoHeadersEnd; i += 2) {
+            HTTP2_NAME_VALIDATOR.validateName(pseudoHeaders[i]);
+            if (pseudoHeaders[i + 1] == null) {
+                throw new IllegalArgumentException("pseudoHeaders value at index " + (i + 1) + " is null");
+            }
+        }
+        final int otherHeadersEnd = otherHeaders.length - 1;
+        for (int i = 0; i < otherHeadersEnd; i += 2) {
+            AsciiString name = otherHeaders[i];
+            HTTP2_NAME_VALIDATOR.validateName(name);
+            if (!seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) != PSEUDO_HEADER_TOKEN) {
+                seenNonPseudoHeader = true;
+            } else if (seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) == PSEUDO_HEADER_TOKEN) {
+                throw new IllegalArgumentException(
```

Comment:
@carl-mastrangelo - Oh I see ... I can clarify the name. It was designed to reduce the byte code for the "hot" path. The though is with validation there will be move overhead incurred anyways so it is not as urgent.